### ISSUE TITLE
Cannot submit to public cloud service using credentials only

### ISSUE THREAD

**Issue by natashadsilva** [IBMStreams/streamsx.topology#2360]
According to 
https://streamsxtopology.readthedocs.io/en/latest/streamsx.topology.context.html#streamsx.topology.context.ConfigParams.SERVICE_DEFINITION

If you submit to the streaming analytics service, you should be able to only specify credentials when you configure job submission, you get this error:

```
Exception in thread "main" java.nio.file.InvalidPathException: Illegal char <"> at index 1: {"streaming-analytics": [{"credentials": {"abc": "def"}, "name": "service"}]}
at sun.nio.fs.WindowsPathParser.normalize(WindowsPathParser.java:194)
at sun.nio.fs.WindowsPathParser.parse(WindowsPathParser.java:165)
at sun.nio.fs.WindowsPathParser.parse(WindowsPathParser.java:89)
at sun.nio.fs.WindowsPath.parse(WindowsPath.java:106)
at sun.nio.fs.WindowsFileSystem.getPath(WindowsFileSystem.java:267)
at java.nio.file.Paths.get(Paths.java:95)
```
Attached a notebook sample.

I tried the second option of 
>A JSON object (dict) of the form: { "type": "streaming-analytics", "name": "service name", "credentials": {...} } with the service credentials as the value of the credentials key.

and also got the same error.

**Comment by natashadsilva** [comment]
[PublicCloudSubmitFailure.zip](https://github.com/IBMStreams/streamsx.topology/files/4314982/PublicCloudSubmitFailure.zip)

**Comment by natashadsilva** [comment]
@markheger

**Comment by RolefH** [comment]
### Reproduction:
**OS running the jupyter-notebook:** linux
**Environment variables:**
unset STREAMING_ANALYTICS_SERVICE_NAME, VCAP_SERVICES, all STREAMS_*


### First option (The service credentials copied from the Service credentials page)
- does *not* work for me
- Notebook code:
```
SA_credentials = getpass.getpass('Streaming Analytics credentials:')
print(type(SA_credentials))    # must be "<class 'str'>"
cfg[ConfigParams.SERVICE_DEFINITION] = SA_credentials
```
Ends up in `TypeError`:
```
TypeError                                 Traceback (most recent call last)
<ipython-input-6-05dffd248a5a> in <module>
----> 1 submission_result = submit_topology(topo)

<ipython-input-4-04db718cb996> in submit_topology(topo)
     20     contextType = context.ContextTypes.STREAMING_ANALYTICS_SERVICE
     21     print(cfg)
---> 22     return context.submit (contextType, topo, config = cfg)

~/miniconda3/envs/streams_py36/lib/python3.6/site-packages/streamsx/topology/context.py in submit(ctxtype, graph, config, username, password)
     72         warnings.warn("Use environment variables STREAMS_USERNAME and STREAMS_PASSWORD", DeprecationWarning, stacklevel=2)
     73 
---> 74     context_submitter = _SubmitContextFactory(graph, config, username, password).get_submit_context(ctxtype)
     75     sr = SubmissionResult(context_submitter.submit())
     76     sr._submitter = context_submitter

~/miniconda3/envs/streams_py36/lib/python3.6/site-packages/streamsx/topology/context.py in get_submit_context(self, ctxtype)
    710             logger.debug("Selecting the STREAMING_ANALYTICS_SERVICE context for submission")
    711             ctxtype = ContextTypes.STREAMING_ANALYTICS_SERVICE
--> 712             return _StreamingAnalyticsSubmitter(ctxtype, self.config, self.graph)
    713         elif ctxtype == 'BUNDLE':
    714             logger.debug("Selecting the BUNDLE context for submission")

~/miniconda3/envs/streams_py36/lib/python3.6/site-packages/streamsx/topology/context.py in __init__(self, ctxtype, config, graph)
    393             # Service name matching the generated VCAP is passed through config.
    394             service_def = self._config().get(ConfigParams.SERVICE_DEFINITION)
--> 395             self._vcap_services = _vcap_from_service_definition(service_def)
    396             self._config()[ConfigParams.SERVICE_NAME] = _name_from_service_definition(service_def)
    397         else:

~/miniconda3/envs/streams_py36/lib/python3.6/site-packages/streamsx/topology/context.py in _vcap_from_service_definition(service_def)
   1553     """
   1554     if 'credentials' in service_def:
-> 1555         credentials = service_def['credentials']
   1556     else:
   1557         credentials = service_def

TypeError: string indices must be integers
```
### Second option (use a JSON object (dict))
- does works for me
- Notebook code:
```
SA_credentials = json.loads(getpass.getpass('Streaming Analytics credentials:'))
print(type(SA_credentials))    # output must be "<class 'dict'>"
cfg[ConfigParams.SERVICE_DEFINITION] = SA_credentials
```
- **But:** In opposite to the documentation, the dict that goes into `cfg[ConfigParams.SERVICE_DEFINITION]` contains the Service credentials only:
```
{
  "apikey": ".........................",
  "iam_apikey_description": "Auto-generated for key ............",
  "iam_apikey_name": "Service credentials-1",
  "iam_role_crn": "...",
  "iam_serviceid_crn": "..................",
  "v2_rest_url": "https://svc-cf.eu-gb................"
}
```

### Second option (use a JSON object (dict)) as interpreted from documentation
**Service credentials *as a JSON string* as value for the `credentials` key** 
- submission fails:
- Notebook code:
```
cfg  = {}
SA_credentials = getpass.getpass('Streaming Analytics credentials:')
service_definition = {}
service_definition['type'] = 'streaming-analytics'
service_definition['name'] = 'Streaming Analytics-GB-lite' # whatever is your service name
service_definition['credentials'] = SA_credentials   # <== a string value
cfg[ConfigParams.SERVICE_DEFINITION] = service_definition
```
- the JSON object for `cfg[ConfigParams.SERVICE_DEFINITION]` is 
```
{
  "credentials": "{   \"apikey\": \"....\",   \"iam_apikey_description\": \"...\",   \"iam_apikey_name\": \"Service credentials-1\",   ...}",
  "name": "Streaming Analytics-GB-lite",
  "type": "streaming-analytics"
}
```
Failure (in the Java submitter):
```
Exception in thread "main" java.lang.IllegalStateException: Not a JSON Object: "{   \"apikey\": \"..............\",   \"iam_apikey_description\": \"Auto-generated ...\",   \"iam_apikey_name\": \"Service credentials-1\",   \"iam_role_crn\": ...}"
at com.google.gson.JsonElement.getAsJsonObject(JsonElement.java:90)
at com.ibm.streamsx.rest.AbstractStreamingAnalyticsService.of(AbstractStreamingAnalyticsService.java:351)
at com.ibm.streamsx.rest.StreamingAnalyticsService.of(StreamingAnalyticsService.java:66)
at com.ibm.streamsx.topology.internal.context.remote.RemoteBuildAndSubmitRemoteContext.streamingAnalyticServiceFromDeploy(RemoteBuildAndSubmitRemoteContext.java:100)
at com.ibm.streamsx.topology.internal.context.remote.RemoteBuildAndSubmitRemoteContext.createSubmissionContext(RemoteBuildAndSubmitRemoteContext.java:47)
at com.ibm.streamsx.topology.internal.context.remote.RemoteBuildAndSubmitRemoteContext.createSubmissionContext(RemoteBuildAndSubmitRemoteContext.java:30)
at com.ibm.streamsx.topology.internal.context.remote.BuildRemoteContext._submit(BuildRemoteContext.java:40)
at com.ibm.streamsx.topology.internal.context.remote.RemoteContextImpl.submit(RemoteContextImpl.java:30)
at com.ibm.streamsx.topology.context.remote.RemoteContextSubmit.main(RemoteContextSubmit.java:38)
```

### Second option (use a JSON object (dict)) as interpreted from documentation
**Service credentials *as a dict value* for the `credentials` key**
- submission succeeds
- Notebook code:
```
cfg  = {}
SA_credentials = json.loads(getpass.getpass('Streaming Analytics credentials:'))
service_definition = {}
service_definition['type'] = 'streaming-analytics'
service_definition['name'] = 'Streaming Analytics-GB-lite'
service_definition['credentials'] = SA_credentials  # <== a dict type
cfg[ConfigParams.SERVICE_DEFINITION] = service_definition
```
- the JSON object for `cfg[ConfigParams.SERVICE_DEFINITION]` is 
```
{
  "credentials": {
    "apikey": ".........",
    "iam_apikey_description": "..........",
    "iam_apikey_name": "Service credentials-1",
    "iam_role_crn": "....",
    "iam_serviceid_crn": "...",
    "v2_rest_url": "https://svc-cf.eu-gb...."
  },
  "name": "Streaming Analytics-GB-lite",
  "type": "streaming-analytics"
}
```

**Comment by RolefH** [comment]
@natashadsilva Seems that you run the notebook on Windows?

**Comment by natashadsilva** [comment]
Yes I am running on Windows.
 
Regards,Natasha D'SilvaIBM Streams DevelopmentIBM Canada - Toronto Lab905-413-5366
ndsilva@ca.ibm.com
 
 
----- Original message -----From: Rolef Heinrich <notifications@github.com>To: "IBMStreams/streamsx.topology" <streamsx.topology@noreply.github.com>Cc: natashadsilva <ndsilva@ca.ibm.com>, Mention <mention@noreply.github.com>Subject: [EXTERNAL] Re: [IBMStreams/streamsx.topology] Cannot submit to public cloud service using credentials only (#2360)Date: Thu, Mar 12, 2020 5:50 AM 
 
@natashadsilva Seems that you run the notebook on Windows?

**Comment by natashadsilva** [comment]
It seems to be a problem with windows.

````
(py36) PS C:\Users\Docker\hostdir> python .\PublicCloudSubmitFailure.py                                                 INFO: streamsx package version: 1.14.10
Submitting on Windows-10-10.0.18362-SP0
Exception in thread "main" java.nio.file.InvalidPathException: Illegal char <"> at index 1: {"streaming-analytics": [{"credentials": {"abc": "def"}, "name": "service"}]}
at sun.nio.fs.WindowsPathParser.normalize(WindowsPathParser.java:194)
at sun.nio.fs.WindowsPathParser.parse(WindowsPathParser.java:165)
at sun.nio.fs.WindowsPathParser.parse(WindowsPathParser.java:89)
at sun.nio.fs.WindowsPath.parse(WindowsPath.java:106)
at sun.nio.fs.WindowsFileSystem.getPath(WindowsFileSystem.java:267)
at java.nio.file.Paths.get(Paths.java:95)
at com.ibm.streamsx.topology.internal.streaminganalytics.VcapServices.getVCAPServices(VcapServices.java:59)
at com.ibm.streamsx.topology.internal.context.remote.RemoteBuildAndSubmitRemoteContext.streamingAnalyticServiceFromDeploy(RemoteBuildAndSubmitRemoteContext.java:99)
at com.ibm.streamsx.topology.internal.context.remote.RemoteBuildAndSubmitRemoteContext.createSubmissionContext(RemoteBuildAndSubmitRemoteContext.java:47)
at com.ibm.streamsx.topology.internal.context.remote.RemoteBuildAndSubmitRemoteContext.createSubmissionContext(RemoteBuildAndSubmitRemoteContext.java:30)
at com.ibm.streamsx.topology.internal.context.remote.BuildRemoteContext._submit(BuildRemoteContext.java:40)
at com.ibm.streamsx.topology.internal.context.remote.RemoteContextImpl.submit(RemoteContextImpl.java:30)
at com.ibm.streamsx.topology.context.remote.RemoteContextSubmit.main(RemoteContextSubmit.java:38)
````

**Comment by natashadsilva** [comment]
On the 4.3.1.1. QSE, running on the same host:
[PublicCloudSubmitFailure.txt](https://github.com/IBMStreams/streamsx.topology/files/4325376/PublicCloudSubmitFailure.txt)


```
[streamsadmin@streamsqse hostdir]$ python PublicCloudSubmitFailure.py 
INFO: streamsx package version: 1.14.10
Submitting on Linux-4.19.76-linuxkit-x86_64-with-centos-7.7.1908-Core
Submitted job
{'_last_modified': 1583892037.11859,
 'json_rep': {'activeViews':...
```

Attached the sample file as python code. Paste credentials from Streams service where indicated to test it.
[PublicCloudSubmitFailure.txt](https://github.com/IBMStreams/streamsx.topology/files/4325394/PublicCloudSubmitFailure.txt)

**Comment by RolefH** [comment]
The interface between the Python and Java is done via environment variable VCAP_SERVICES, which Python sets when it creates the JVM to invoke the submission. The environment variable contains the VCAP JSON string 
```
{"streaming-analytics": [{"credentials": {"apikey": "...", ...}, ...}]}
```
For other ways to configure the cloud service, the environment variable VCAP_SERVICES can contain the filename of a JSON file with vcap service information.

The Java part tries to test if the value of the VCAP_SERVICES environment variable is a regular file. If yes, the file is read, assuming it contains valid vcap information. If it is not a regular file, the variable content is taken 'as is'.

Testing a JSON string for being a regular file fails in Windows with the observed exception, but succeeds on Linux.

Following code succeeds on Linux, but fails on Windows:
```
package test;
import java.nio.file.Files;
import java.nio.file.Paths;

public class Main {
    public static void main(String[] args) {
        String testString = "{\"streaming-analytics\": ["
                + "{\"credentials\": {\"apikey\": \"A\", "
                + "\"iam_apikey_description\": \"B\", \"iam_apikey_name\": \"C\", "
                + "\"iam_role_crn\": \"D\", \"iam_serviceid_crn\": \"E\", "
                + "\"v2_rest_url\": \"https://blah.com\"}, "
                + "\"name\": \"service\"}]}";
        if (Files.isRegularFile(Paths.get(testString))) {    // Paths.get(...) throws Exc in Windows
            System.out.println ("Reading file ...");
        } else {
            System.out.println ("Using string 'as is' ...");
        }
    }
}
```
**Run it on Linux:**
```
Using string 'as is' ...
```
**Run it on Windows:**
```
Exception in thread "main" java.nio.file.InvalidPathException: Illegal char <"> at index 1: {"streaming-analytics": [{"credentials": {"apikey": "A", "iam_apikey_description": "B", "iam_apikey_name": "C", "iam_role_crn": "D", "iam_serviceid_crn": "E", "v2_rest_url": "https://blah.com"}, "name": "service"}]}
        at java.base/sun.nio.fs.WindowsPathParser.normalize(WindowsPathParser.java:182)
        at java.base/sun.nio.fs.WindowsPathParser.parse(WindowsPathParser.java:153)
        at java.base/sun.nio.fs.WindowsPathParser.parse(WindowsPathParser.java:77)
        at java.base/sun.nio.fs.WindowsPath.parse(WindowsPath.java:92)
        at java.base/sun.nio.fs.WindowsFileSystem.getPath(WindowsFileSystem.java:229)
        at java.base/java.nio.file.Paths.get(Paths.java:84)
        at test.Main.main(Main.java:14)
```

Using `Paths.get(...)` is causing problems in Windows.

**Comment by RolefH** [comment]
resolved with [1.14.11 release](https://github.com/IBMStreams/streamsx.topology/releases/tag/v1.14.11)
